rclcpp/include/rclcpp/wait_for_message.hpp

```diff
@@ -0,0 +1,81 @@
+// Copyright 2021 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCLCPP__WAIT_FOR_MESSAGE_HPP_
+#define RCLCPP__WAIT_FOR_MESSAGE_HPP_
+
+#include <memory>
+#include <string>
+
+#include "rclcpp/node.hpp"
+#include "rclcpp/visibility_control.hpp"
+#include "rclcpp/wait_set.hpp"
+
+namespace rclcpp
+{
+/// Wait for the next incoming message.
+/**
+ * Given an already initialized subscription,
+ * wait for the next incoming message to arrive before the specified timeout.
+ *
+ * \param[out] out is the message to be filled when a new message is arriving.
+ * \param[in] subscription shared pointer to a previously initialized subscription.
+ * \param[in] time_to_wait parameter specifying the timeout before returning.
+ * \return true if a message was successfully received, false otherwise.
+ */
+template<class MsgT, class Rep = int64_t, class Period = std::milli>
+bool wait_for_message(
+  MsgT & out,
+  std::shared_ptr<rclcpp::Subscription<MsgT>> subscription,
+  std::chrono::duration<Rep, Period> time_to_wait = std::chrono::duration<Rep, Period>(-1))
```

Comment:
One issue here is that the `rclcpp::WaitSet` does not add a guard condition that is triggered on shutdown. So if you use `-1` this will actually block forever, until a message is received, even if you get a SIGINT or shutdown is called somewhere.

So you'd either want to add that yourself, or use an executor. I know you want to take the message without a callback/callback_group, so the wait set is more ideal. But you'll need to add the guard condition yourself. You can see how executors do it:

https://github.com/ros2/rclcpp/blob/e9e398d2d49de2ecc3148d425f8659d9db74918c/rclcpp/src/rclcpp/executor.cpp#L59-L65

Reply:
I've added a guard condition and hooked it up to the shutdown callbacks of the context. I am deducing the context from the incoming node - hope this is the right approach.

I couldn't really find another way of detecting whether it was indeed the GC which triggered the waitset to return so I've manually added a `shutdown_requested` flag. Let me know if there's a better way for it.